### PR TITLE
Added coffee-script cheat sheet

### PR DESCRIPTION
**Pull request by codenirvana**
**What does your Instant Answer do?**
My cheat sheet provides help for [CoffeeScript](http://coffeescript.org)

**What problem does your Instant Answer solve (Why is it better than organic links)?**
Provides a cheat sheet for common CoffeeScript syntax and  usage without visiting the main website or checking any manual.

**What is the data source for your Instant Answer? (Provide a link if possible)**
The CoffeeScript website and cheatography.com

**Why did you choose this data source?**
It's the source.

**Are there any other alternative (better) data sources?**
Not that I'm aware of.

**What are some example queries that trigger this Instant Answer?**
coffeescript cheat sheet, coffeescript  Installation, coffeescript operators 

**Which communities will this Instant Answer be especially useful for? (gamers, book lovers, etc)**
Programmers, JavaScript developers

**Is this Instant Answer connected to a DuckDuckHack Instant Answer idea?**
Yes

**Which existing Instant Answers will this one supersede/overlap with?**
N/A

**Are you having any problems? Do you need our help with anything?**
No

**What are the terms of use for the API? Will DuckDuckGo need specific authorization (e.g. an API key)? Are there any costs associated with API usage?**
No

**Where did you hear about DuckDuckHack? (For first time contributors)**
DuckDuckGo

**What does the Instant Answer look like? (Provide a screenshot for new or updated Instant Answers)**
coffeescript-cheat-sheet

## Checklist
Please place an 'X' where appropriate.

```
[X] Added metadata and attribution information
[] Wrote test file and added to t/ directory
[X] Verified that instant answer adheres to design guidelines (https://duck.co/duckduckhack/design_styleguide)
[X] Verified that instant answer adheres to code styleguide (https://duck.co/duckduckhack/code_styleguide)
[] Tested cross-browser compatibility

    Please let us know which browsers/devices you've tested on:
    - Windows 8
        [] Google Chrome
        [] Firefox
        [] Opera
        [] IE 10

    - Windows 7
        [] Google Chrome
        [] Firefox
        [] Opera
        [] IE 8
        [] IE 9
        [] IE 10

    - Windows XP
        [] IE 8

    - Mac OSX
        [] Google Chrome
        [] Firefox
        [] Opera
        [] Safari

    - iOS (iPhone)
        [] Safari Mobile
        [] Google Chrome
        [] Opera

    - iOS (iPad)
        [] Safari Mobile
        [] Google Chrome
        [] Opera

    - Android
        [] Firefox
        [] Native Browser
        [] Google Chrome
        [] Opera
```

------
IA Page: https://duck.co/ia/view/coffee_script_cheat_sheet

### CODE REVIEW

**Comment by moollaza** [comment]
@codenirvana thanks a lot for contributing, we really appreciate it! We'll be sure to review this shortly and give you some feedback!

> Don't know how to generate a screenshot

@codenirvana did you follow our environment setup instructions? If you're using Codio you will have DuckPAN pre-installed. This will let you run the code and preview your new Instant Answer. I would suggest you do that and run `duckpan server` so you can preview this IA

**Comment by codenirvana** [comment]
@moollaza thanks DuckPAN worked!
But my cheat sheet is not working, getting this error:
> Sorry, no hit for your instant answer

What's wrong?

**Comment by moollaza** [comment]
@codenirvana was your query "coffee script cheat sheet" or "coffeescript cheat sheet"? Right now the triggering is based on the filename (we'll soon allow you to define additional triggers) so in this case it will trigger on `coffee script` but not `coffeescript`

File: share/goodie/cheat_sheets/coffee-script.json [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    "section_order": [
+        "Installation",
+        "Usage",
+        "Variables",
+        "Functions",
+        "Arrays",
+        "Operators and aliases"
+    ],
+    "sections": {
+        "Installation": [{
```
Comment (moollaza) [inline share/goodie/cheat_sheets/coffee-script.json]: I'm not sure if this one is necessary. We have the NPM Spice which shows how to install NPM modules.

File: share/goodie/cheat_sheets/coffee-script.json [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+            "key": "array\\[1..3\\] = \\[3, 2, 1\\]"
+        }, {
+            "val": "Reverse elements in array",
+            "key": "ground_coffee.reverse()"
+        }, {
+            "val": "Max element of array",
+            "key": "Math.max \\[12, 32, 11, 67, 1, 3\\]..."
+        }, {
+            "val": "Concatenating arrays",
+            "key": "[1 ,2, 3].concat \\[ 4, 5, 6\\]"
```
Comment (moollaza) [inline share/goodie/cheat_sheets/coffee-script.json]: Looks like the first 2 brackets need to be escaped

File: share/goodie/cheat_sheets/coffee-script.json (hunk 1 above)
Comment (codenirvana) [inline share/goodie/cheat_sheets/coffee-script.json]: Right, I will remove installation step.

**Comment by codenirvana** [comment]
@moollaza fixed the issue(added screenshot).
Renamed file, now its working for "coffeescript cheat sheet".

**Comment by mintsoft** [comment]
@codenirvana if you did want to make it work for bother `coffeescript` and `coffee script` you can commit in a symlink with the alternative name

**Comment by codenirvana** [comment]
@mintsoft thanks its working for both now.

File: share/goodie/cheat_sheets/coffeescript.json [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+            "key": "array1 = array\\[1..3\\]"
+        }, {
+            "val": "Change three items of array",
+            "key": "array\\[1..3\\] = \\[3, 2, 1\\]"
+        }, {
+            "val": "Reverse elements in array",
+            "key": "ground_coffee.reverse()"
+        }, {
+            "val": "Max element of array",
+            "key": "Math.max \\[12, 32, 11, 67, 1, 3\\]..."
```
Comment (moollaza) [inline share/goodie/cheat_sheets/coffeescript.json]: In an effort to be a little more succinct, let's change this to: `Math.max [12, 32, ...]`

File: share/goodie/cheat_sheets/coffeescript.json [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        }, {
+            "val": "Concatenating arrays",
+            "key": "\\[1 ,2, 3\\].concat \\[ 4, 5, 6\\]"
```
Comment (moollaza) [inline share/goodie/cheat_sheets/coffeescript.json]: Similarly, let's try: `[1, 2].concat [4, 5, 6]`

Looks like there's also a space after the `1`, before the `,`

**Comment by codenirvana** [comment]
@moollaza fixed spacing issues.  Anything else?

File: share/goodie/cheat_sheets/coffeescript.json [hunk 5]
```diff
@@ -0,0 +1,126 @@
+{
+    "id": "coffeescript_cheat_sheet",
```
Comment (moollaza) [inline share/goodie/cheat_sheets/coffeescript.json]: @codenirvana one last fix! Please change to `coffee_script_cheat_sheet`

File: share/goodie/cheat_sheets/coffeescript.json [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+        }],
+        "Arrays": [{
+            "val": "Array with numbers from 1 to 4",
+            "key": "array = \\[1, 2, 3, 4\\]"
+        }, {
+            "val": "Array with numbers from 1 to 4",
+            "key": "array = \\[1...4\\]"
+        }, {
+            "val": "Array with numbers from 1 to 4",
+            "key": "array = \\[1....3\\]"
```
Comment (moollaza) [inline share/goodie/cheat_sheets/coffeescript.json]: Looks like you have 4 `.` here, and 3 before and 2 below. Let's normalize to 3? `...`

**Comment by moollaza** [comment]
@codenirvana juuuust about there. Just need to fix the ID and then normalize our usage of `.`. After that this is good for merge.

Thanks again for contributing and sticking around to fix everything. We really appreciate it!

**Comment by codenirvana** [comment]
@moollaza fixed the syntax errors and cross checked by compiling the coffeescript code.
```..``` & ```...``` is according to its syntax.

**Comment by moollaza** [comment]
> .. & ... is according to its syntax.

Aah. Thanks for clarifying. This :+1:

**Comment by moollaza** [comment]
@codenirvana Thanks again for taking the time to contribute! We really appreciate it. Feel free to stick around and comment/help on other PR's & Issues or even submit more Instant Answers!

**Comment by codenirvana** [comment]
@moollaza sure, I will submit more IAs soon!
I have a problem(off topic) at **Community Platform**: When I check the box *Show IAs of which I am the developer* it shows nothing. Please tell me how to link my IAs with *Community Platform*.
Also how to change the **DEVELOPER** link of my IA Pages to [Udit Vasu](https://duck.co/user/codenirvana)

**Comment by moollaza** [comment]
@codenirvana congrats! It's live: https://duckduckgo.com/?q=coffee%20script%20cheat%20sheet&ia=cheatsheet

Thanks again for taking the time to contribute! We really appreciate it. Feel free to stick around and comment/help on other PR's, tackle open issues or even submit more Instant Answers!

------

> Please tell me how to link my IAs with Community Platform.

We're currently rolling out Instant Answer page access to DuckDuckHack developers which will allow them to edit the IA pages for their in-development Instant Answers.

If you’d like to have edit permission for your IA Page(s) please make a duck.co account, then link your GitHub account to it, and finally send me your duck.co account name and a list of the in-progress IA’s you’re working on. I'll get the team to give you edit permissions so you can fill out the IA pages.

> Also how to change the DEVELOPER link of my IA Pages to Udit Vasu

Currently we only Duck.co account (or GitHub for users without Ducko.co accounts). You can add links to your other social profiles and websites on your Duck.co account if you'd like!

**Comment by codenirvana** [comment]
@moollaza Already linked my duck.co account with GitHub. My username is [codenirvana](https://duck.co/user/codenirvana) .

Also I seen that you update my IA pages link, Thanks!
Can you please change the link text to my name instead of username, that will look nice :+1:

**Comment by moollaza** [comment]
@codenirvana everything should be updated! Please let me know if I missed anything

**Comment by codenirvana** [comment]
@moollaza That's great.
I was just thinking to have contributor 'name' as the link instead of 'username'

**Pull request merged**